src/journal/journald-socket.c

```diff
@@ -0,0 +1,107 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <sys/socket.h>
+#include <sys/uio.h>
+
+#include "fd-util.h"
+#include "iovec-util.h"
+#include "journald-socket.h"
+#include "log.h"
+#include "macro.h"
+#include "process-util.h"
+#include "socket-util.h"
+
+void server_open_forward_socket(Server *s) {
+        _cleanup_close_ int socket_fd = -EBADF;
+        const SocketAddress *addr;
+        int family;
+
+        assert(s);
+        assert(s->forward_socket_fd < 0);
+
+        addr = &s->forward_to_socket;
+
+        family = socket_address_family(addr);
+
+        if (!IN_SET(family, AF_UNIX, AF_INET, AF_INET6, AF_VSOCK)) {
+                log_debug("Unsupported socket type for forward socket: %d", family);
+                return;
+        }
+
+        socket_fd = socket(family, SOCK_STREAM|SOCK_CLOEXEC, 0);
+        if (socket_fd < 0) {
+                log_debug_errno(errno, "Failed to create forward socket, ignoring: %m");
+                return;
+        }
+
+        if (connect(socket_fd, &addr->sockaddr.sa, addr->size) < 0) {
+                log_debug_errno(errno, "Failed to connect to remote address for forwarding, ignoring: %m");
+                return;
+        }
+
+        s->forward_socket_fd = TAKE_FD(socket_fd);
+        log_debug("Successfully connected to remote address for forwarding");
+}
+
+void server_forward_socket(
+                Server *s,
+                const struct iovec *iovec,
+                size_t n_iovec,
+                int priority) {
+        _cleanup_free_ struct iovec *iov_alloc = NULL;
+        struct iovec *iov = NULL;
+
+        assert(s);
+        assert(iovec);
+        assert(n_iovec > 0);
+
+        if (LOG_PRI(priority) > s->max_level_socket)
+                return;
+
+        /* only open a forwarding socket if a forwarding address has been set
+         * and we are in the main namespace and the forwarding socket is not already open */
+        if (s->forward_to_socket.sockaddr.sa.sa_family != AF_UNSPEC && !s->namespace && s->forward_socket_fd < 0)
```

Comment:
please move this conditionalization into server_open_forward_socket() itself, i.e. make it a clean NOP if there's nothing to do.